design/one-pager-package-image-config.md

```diff
@@ -0,0 +1,377 @@
+Controller only operates on Pods or the native Kubernetes resources resulting in
+a Pod, e.g. Deployments, StatefulSets, etc. There is no way to configure it to
+work with Crossplane packages.
+
+Kyverno, on the other hand, is a general-purpose policy engine for Kubernetes
+that allows users to define policies for any Kubernetes resource. Using its
+`ClusterPolicy` API, one can define [rules to verify images] and by leveraging
+the `imageExtractors` field, it is possible to configure it to work with
+Crossplane packages. See the [using Kyverno for package image verification]
+section for more details.
+
+Another example in this space is the project [Flux]. Compared to the previous
+APIs, this project provides a [minimalistic API to verify images] with the
+`OCIRepository` API, marking some parts of the API (e.g. keyless verification)
+as experimental.
+
+In Crossplane itself, we had [a previous proposal] aiming to solve signature
+verification for packages but mostly focusing on the implementation details,
+primarily for verification using a shared public key. The follow-up
+[implementation PR] extended the solution also to support keyless verification
+by mostly following the same solution / similar API as Flux by extending the
+Package installation APIs for verification settings, leaving the dependencies
+out of the scope. This proposal aims to provide a more comprehensive solution by
+introducing an API that can be extended to support various verification settings
+and also covering the ones coming as a dependency.
+
+## Goals
+
+- Allow configuration of credentials for package images and their dependencies
+  at runtime.
+- Enable users to set policies for verifying image signatures.
+- Define an image-centric configuration API that centralizes settings for
+  package images.
+
+## Proposal
+
+This document proposes a new API, `ImageConfig`, under the `pkg.crossplane.io`
+API group, that allows users to configure settings for package images. The API
+enables users to define rules for matching images and configuring how to
+interact with the registries hosting the images, including authentication and
+TLS settings. It also provides a way to define policies for verifying image
+signatures.
+
+Considering the API is solving some existing problems, e.g. configuring pull
+secrets for dependencies, and not introducing a behavioral change unless the
+users opt in by creating an `ImageConfig` object, we believe this API should be
+introduced as a beta API without going through the feature lifecycle stages.
+
+### API
+
+```yaml
+apiVersion: pkg.crossplane.io/v1beta1
+kind: ImageConfig
+metadata:
+  name: acme-packages
+spec:
+  matchImages:
+    - prefix: registry1.com/acme-co/configuration-foo
+    - prefix: registry1.com/acme-co/configuration-bar
+    - prefix: registry1.com/acme-co/function-baz
+  registry:
+    authentication:
+      pullSecretRef:
+        name: acme-registry-credentials
```

Comment:
How does this secret interact with the pull secrets you can specify "directly" when you install a package? Does it try both?

Reply:
Yes, the auth chain method we use accepts multiple image pull secrets. This will be additive to that list and similarly to runtime deployments `imagePullSecrets` list.

Reply:
Could you update the design to mention that explicitly please. (If it's not already there - I didn't notice it.)